Add an `:examples/core-hello-world` project

zen-dog opened the pull request:
Summary:
Run the hello-world example framework that:
- relies on our `core` module to handle Mesos connection, event handling, offer matching etc. as
  opposed to the `simple-hello-world` example framework which uses only the Mesos client and has to
  implement all of the above-mentioned pieces itself
- starts one `echo "Hello, world" && sleep N` task
- exits when the task fails (or finishes)

JIRA: DCOS-47480

zen-dog committed:
Add an `:examples/core-hello-world` project

Summary:
Run the hello-world example framework that:
- relies on our `core` module to handle Mesos connection, event handling, offer matching etc. as
  opposed to the `simple-hello-world` example framework which uses only the Mesos client and has to
  implement all of the above mentioned pieces itself
- starts one `echo "Hello, world" && sleep N` task
- exits should the task fail (or fail to start)

JIRA: DCOS-47480

zen-dog committed:
Add the README file

zen-dog committed:
Simplified the example by using `Source.maybe` instead of `Source.queue`

zen-dog commented:
There are a few issues to resolve before merging:
- we need to ACK Mesos status updates
- there seems to be a bug where we accept offers where there is nothing to launch
- core-hello-world needs an integration test

But otherwise, it looks good

zen-dog committed:
Update `application.conf` with `akka.http` settings

zen-dog committed:
Renamed main example class to `CoreHelloWorldFramework`

jeschkies committed:
Replace tab.

jeschkies commented on examples/core-hello-world/src/main/resources/logback.xml:
```diff
@@ -0,0 +1,18 @@
+    <root level="DEBUG">
+        <appender-ref ref="stdout"/>
+    </root>
+</configuration>
```
I think we should add the configuration for structured logging. What do you think, @takirala?

takirala replied:
👍 we have one example at https://github.com/mesosphere/usi/blob/master/examples/hello-world/src/main/resources/logback.groovy

alenkacz replied:
yeah, let's go back to what @takirala did in that link, please :)

jeschkies commented on examples/core-hello-world/src/main/scala/com/mesosphere/usi/helloworld/CoreHelloWorld.scala:
```diff
@@ -0,0 +1,166 @@
+    .setFailoverTimeout(0d)
+    .build()
+
+  val client: MesosClient = Await.result(MesosClient(settings, frameworkInfo).runWith(Sink.head), 10.seconds)
```
Hm, I wonder if we really should provide a constructed client or if we should just inject an implementation. What do you think, @timcharper?

jeschkies commented on examples/core-hello-world/src/main/scala/com/mesosphere/usi/helloworld/CoreHelloWorld.scala:
```diff
@@ -0,0 +1,166 @@
+    // state events. This makes the event handling a simple match-case
+    .flatMapConcat {
+      case (snapshot, updates) =>
+        updates.prepend(Source.single(snapshot))
```
I don't really get this. Why isn't the scheduler flow prepending for us, @timcharper  and @alenkacz?

jeschkies reviewed (changes requested):
It feels this framework is kind of cheating. It does not tell me how I would send `PodUpdated` events through the scheduler interface. That seems to be the most important thing of our architecture.

alenkacz commented:
@zen-dog 

> There are a few issues to resolve before merging:
> 
> * we need to ACK Mesos status updates
> * there seems to be a bug where we accept offers where there is nothing to launch
> * core-hello-world needs an integration test
> 
> But otherwise, it looks good

ACKing is done here #50
Decline was implemented in #49

zen-dog committed:
Rebased on master

- `CoreHelloWorldFrameworkTest` in now green and also tests that the scheduler produces a `PodSpecUpdated` event
- the framework itself is now materialized to a `Sink.queue` with `StateEvent`s for easier testing

zen-dog committed:
Correct formatting

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/logic/MesosEventsLogic.scala:
```diff
@@ -59,7 +59,8 @@ private[core] class MesosEventsLogic(mesosCallFactory: MesosCalls) extends Impli
                 taskId.asProto,
-                name = "hi",
+                // we use sanitized podId as the task name for now
+                name = podSpec.id.value.replaceAll("[^a-zA-Z0-9-]", ""),
```
:D

alenkacz commented on examples/core-hello-world/README.md:
```diff
@@ -0,0 +1,8 @@
+  opposed to the `simple-hello-world` example framework which uses only the Mesos client and has to 
+  implement all of the above mentioned pieces itself
+- starts one `echo "Hello, world" && sleep N` task
+- exits should the task fail (or fail to start)
```
I don't understand. Is this sentence correct? are you saying it exits if the task fail?

zen-dog replied:
Yep, it does.

alenkacz commented on examples/core-hello-world/src/main/scala/com/mesosphere/usi/examples/CoreHelloWorldFramework.scala:
```diff
@@ -0,0 +1,164 @@
+  *    opposed to the `simple-hello-world` framework which uses only the Mesos client and has to implement
+  *    all of the above mentioned pieces itself
+  *  - starts one `echo "Hello, world" && sleep N` task
+  *  - exits should the task fail (or fail to start)
```
if you fix this sentence there, do it here as well :) thanks

alenkacz commented on examples/core-hello-world/src/main/scala/com/mesosphere/usi/examples/CoreHelloWorldFramework.scala:
```diff
@@ -0,0 +1,164 @@
+    * +--------------------+         +-----------------------+        +-----------------------+
+    *
+    *
+    * In a nutshell: frameworks sends [[PodSpec]] updates down the flow and receives [[StateEvent]]s when things
```
nice docs 👍

alenkacz reviewed (approved):
No blockers, nice docs 👍  

only it would be nice to pull in the structured logging settings Tarun prepared, please.

zen-dog commented:
> only it would be nice to pull in the structured logging settings Tarun prepared, please.

I thought about that and decided not too. While structured logging is good for elk-style log analysis, when playing with those examples locally it is much clearer *not* to have it ;)

zen-dog committed:
Minor documentation improvement
